query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java

```diff
@@ -79,31 +81,32 @@ private Object extractValue(Object wrappedValue) {
    @Override
    @SuppressWarnings("unchecked")
    public Void call() throws Exception {
-      preIndex();
-      KeyValueFilter filter = getFilter();
-      try (Stream<CacheEntry<Object, Object>> stream = cache.getAdvancedCache().withFlags(Flag.CACHE_MODE_LOCAL)
-              .cacheEntrySet().stream()) {
-         Iterator<CacheEntry<Object, Object>> iterator = stream.filter(CacheFilters.predicate(filter)).iterator();
-         while (iterator.hasNext()) {
-            CacheEntry<Object, Object> next = iterator.next();
-            Object value = extractValue(next.getValue());
-            if (value != null && value.getClass().equals(entity))
-               indexUpdater.updateIndex(next.getKey(), value);
+      if (keys == null) {
+         preIndex();
+         KeyValueFilter filter = getFilter();
+         try (Stream<CacheEntry<Object, Object>> stream = cache.getAdvancedCache().withFlags(Flag.CACHE_MODE_LOCAL)
+               .cacheEntrySet().stream()) {
+            Iterator<CacheEntry<Object, Object>> iterator = stream.filter(CacheFilters.predicate(filter)).iterator();
+            while (iterator.hasNext()) {
+               CacheEntry<Object, Object> next = iterator.next();
+               Object value = extractValue(next.getValue());
+               if (value != null && value.getClass().equals(entity))
+                  indexUpdater.updateIndex(next.getKey(), value);
+            }
+         }
+         postIndex();
+      } else {
+         for (Object key : keys) {
+            Object value = extractValue(cache.get(key));
+            if (value != null) {
+               indexUpdater.updateIndex(key, value);
+               indexUpdater.flush(value.getClass());
```

Comment:
Since flush can be potentially an expensive operation, better to do it after all the keys are indexed.